File: apple/ConduitKit/Sources/ConduitProtocol/Framing.swift
import Foundation

/// TLV framing over the byte stream (spec §5.4): 1-byte kind, 4-byte big-endian
/// length, payload. Control frames carry canonical JSON; file-chunk frames carry
/// a fixed 25-byte binary header plus raw data (bulk side-channel, spec §6).
public enum FrameKind: UInt8, Sendable {
    case control = 0x01
    case fileChunk = 0x02
    case screenFrame = 0x03
}

public struct ChunkFrame: Sendable, Equatable {
    public var fileID: UUID
    public var seq: UInt64
    public var isLast: Bool
    public var data: Data

    public init(fileID: UUID, seq: UInt64, isLast: Bool, data: Data) {
        self.fileID = fileID
        self.seq = seq
        self.isLast = isLast
        self.data = data
    }
}

/// A single encoded video frame on the wire (spec §6 SCREEN_FRAME, Phase 3).
/// Binary bulk data on the dedicated screen connection, parallel to ChunkFrame.
/// `sessionID` is a small per-connection id (multiple screen sessions never
/// share one bulk connection in v1, but the field keeps the format future-proof).
public struct ScreenFrame: Sendable, Equatable {
    public var sessionID: UInt16
    public var seq: UInt32
    public var isKeyframe: Bool
    /// Presentation timestamp, milliseconds; monotonic per session.
    public var ptsMillis: UInt64
    /// Packed encoded frame (parameter sets + sample data); see ScreenFrameCodec.
    public var data: Data

    public init(sessionID: UInt16, seq: UInt32, isKeyframe: Bool, ptsMillis: UInt64, data: Data) {
        self.sessionID = sessionID
        self.seq = seq
        self.isKeyframe = isKeyframe
        self.ptsMillis = ptsMillis
        self.data = data
    }
}

public enum Frame: Sendable, Equatable {
    case control(Data)
    case fileChunk(ChunkFrame)
    case screenFrame(ScreenFrame)
}

public enum FramingError: Error, Equatable {
    case oversizedFrame(kind: UInt8, length: Int)
    case malformedChunk
    case malformedScreenFrame
}

public enum FrameCodec {
    public static func encodeControl(_ json: Data) -> Data {
        var out = Data(capacity: 5 + json.count)
        out.append(FrameKind.control.rawValue)
        out.appendBigEndian(UInt32(json.count))
        out.append(json)
        return out
    }

    public static func encode(_ chunk: ChunkFrame) -> Data {
        var out = Data(capacity: 5 + ProtocolConstants.chunkHeaderSize + chunk.data.count)
        out.append(FrameKind.fileChunk.rawValue)
        out.appendBigEndian(UInt32(ProtocolConstants.chunkHeaderSize + chunk.data.count))
        withUnsafeBytes(of: chunk.fileID.uuid) { out.append(contentsOf: $0) }
        out.appendBigEndian(chunk.seq)
        out.append(chunk.isLast ? 1 : 0)
        out.append(chunk.data)
        return out
    }

    static func decodeChunkPayload(_ payload: Data) throws -> ChunkFrame {
        guard payload.count >= ProtocolConstants.chunkHeaderSize else {
            throw FramingError.malformedChunk
        }
        let bytes = [UInt8](payload)
        let uuid = NSUUID(uuidBytes: bytes) as UUID
        var seq: UInt64 = 0
        for i in 16..<24 {
            seq = (seq << 8) | UInt64(bytes[i])
        }
        let isLast = bytes[24] != 0
        let data = Data(bytes[ProtocolConstants.chunkHeaderSize...])
        return ChunkFrame(fileID: uuid, seq: seq, isLast: isLast, data: data)
    }

    public static func encode(_ frame: ScreenFrame) -> Data {
        var out = Data(capacity: 5 + ProtocolConstants.screenFrameHeaderSize + frame.data.count)
        out.append(FrameKind.screenFrame.rawValue)
        out.appendBigEndian(UInt32(ProtocolConstants.screenFrameHeaderSize + frame.data.count))
        out.appendBigEndian(frame.sessionID)
        out.appendBigEndian(frame.seq)
        out.append(frame.isKeyframe ? 1 : 0)
        out.appendBigEndian(frame.ptsMillis)
        out.append(frame.data)
        return out
    }

    static func decodeScreenPayload(_ payload: Data) throws -> ScreenFrame {
        guard payload.count >= ProtocolConstants.screenFrameHeaderSize else {
            throw FramingError.malformedScreenFrame
        }
        let bytes = [UInt8](payload)
        let sessionID = UInt16(bytes[0]) << 8 | UInt16(bytes[1])
        var seq: UInt32 = 0
        for i in 2..<6 { seq = (seq << 8) | UInt32(bytes[i]) }
        let isKeyframe = bytes[6] != 0
        var pts: UInt64 = 0
        for i in 7..<15 { pts = (pts << 8) | UInt64(bytes[i]) }
        let data = Data(bytes[ProtocolConstants.screenFrameHeaderSize...])
        return ScreenFrame(sessionID: sessionID, seq: seq, isKeyframe: isKeyframe, ptsMillis: pts, data: data)
    }
}

/// Incremental frame parser. Feed it arbitrary byte segments from the stream;
/// it yields complete frames. Frames with an unknown kind byte are skipped
/// (forward compatibility), matching the unknown-type rule for messages.
public struct FrameReader: Sendable {
    private var buffer = Data()
    /// Count of skipped unknown-kind frames, surfaced for the debug HUD/logs.
    public private(set) var skippedUnknownFrames = 0

    public init() {}

    public mutating func append(_ data: Data) throws -> [Frame] {
        buffer.append(data)
        var frames: [Frame] = []
        while true {
            guard buffer.count >= 5 else { break }
            let bytes = buffer.prefix(5)
            let kindByte = bytes[bytes.startIndex]
            var length: UInt32 = 0
            for i in 1...4 {
                length = (length << 8) | UInt32(bytes[bytes.startIndex + i])
            }
            let payloadLength = Int(length)
            let maxAllowed = max(
                ProtocolConstants.maxControlPayload,
                max(
                    ProtocolConstants.maxChunkData + ProtocolConstants.chunkHeaderSize,
                    ProtocolConstants.maxScreenFrameData + ProtocolConstants.screenFrameHeaderSize
                )
            )
            guard payloadLength <= maxAllowed else {
                throw FramingError.oversizedFrame(kind: kindByte, length: payloadLength)
            }
            guard buffer.count >= 5 + payloadLength else { break }
            let payload = Data(buffer.dropFirst(5).prefix(payloadLength))
            buffer = Data(buffer.dropFirst(5 + payloadLength))

            // The `maxAllowed` check above is only the shared ceiling — it lets a
            // file chunk ride in at the *screen* frame's 4 MiB limit, twice the
            // 2 MiB cap the protocol documents for chunks. Each kind is held to
            // its own cap here (2026-08-17; Go and Kotlin do the same).
            switch FrameKind(rawValue: kindByte) {
            case .control:
                guard payload.count <= ProtocolConstants.maxControlPayload else {
                    throw FramingError.oversizedFrame(kind: kindByte, length: payloadLength)
                }
                frames.append(.control(payload))
            case .fileChunk:
                guard payloadLength <= ProtocolConstants.maxChunkData + ProtocolConstants.chunkHeaderSize else {
                    throw FramingError.oversizedFrame(kind: kindByte, length: payloadLength)
                }
                frames.append(.fileChunk(try FrameCodec.decodeChunkPayload(payload)))
            case .screenFrame:
                guard payloadLength <= ProtocolConstants.maxScreenFrameData + ProtocolConstants.screenFrameHeaderSize else {
                    throw FramingError.oversizedFrame(kind: kindByte, length: payloadLength)
                }
                frames.append(.screenFrame(try FrameCodec.decodeScreenPayload(payload)))
            case nil:
                skippedUnknownFrames += 1
            }
        }
        return frames
    }
}

extension Data {
    mutating func appendBigEndian(_ value: UInt16) {
        Swift.withUnsafeBytes(of: value.bigEndian) { append(contentsOf: $0) }
    }

    mutating func appendBigEndian(_ value: UInt32) {
        Swift.withUnsafeBytes(of: value.bigEndian) { append(contentsOf: $0) }
    }

    mutating func appendBigEndian(_ value: UInt64) {
        Swift.withUnsafeBytes(of: value.bigEndian) { append(contentsOf: $0) }
    }
}

File: core/wire/framing.go
package wire

import (
	"encoding/binary"
	"errors"
)

// TLV framing (spec §5.4): kind u8, length u32be, payload.
const (
	KindControl     byte = 0x01
	KindFileChunk   byte = 0x02
	KindScreenFrame byte = 0x03

	chunkHeaderSize       = 25 // uuid(16) + seq(8) + flags(1)
	screenFrameHeaderSize = 15 // sessionId(2) + seq(4) + flags(1) + pts(8)

	MaxControlPayload = 1 << 20
	MaxChunkData      = 2 << 20
	MaxScreenData     = 4 << 20
)

var (
	ErrOversizedFrame  = errors.New("oversized frame")
	ErrMalformedChunk  = errors.New("malformed chunk frame")
	ErrMalformedScreen = errors.New("malformed screen frame")
)

// ChunkFrame is a file-transfer bulk frame (kind 0x02).
type ChunkFrame struct {
	FileID [16]byte
	Seq    uint64
	IsLast bool
	Data   []byte
}

// ScreenFrame is an encoded video frame (kind 0x03).
type ScreenFrame struct {
	SessionID  uint16
	Seq        uint32
	IsKeyframe bool
	PtsMillis  uint64
	Data       []byte
}

// Frame is one parsed frame: exactly one field is non-nil / set by Kind.
type Frame struct {
	Kind    byte
	Control []byte
	Chunk   *ChunkFrame
	Screen  *ScreenFrame
}

func EncodeControl(json []byte) []byte {
	out := make([]byte, 5+len(json))
	out[0] = KindControl
	binary.BigEndian.PutUint32(out[1:5], uint32(len(json)))
	copy(out[5:], json)
	return out
}

func EncodeChunk(c ChunkFrame) []byte {
	payloadLen := chunkHeaderSize + len(c.Data)
	out := make([]byte, 5+payloadLen)
	out[0] = KindFileChunk
	binary.BigEndian.PutUint32(out[1:5], uint32(payloadLen))
	copy(out[5:21], c.FileID[:])
	binary.BigEndian.PutUint64(out[21:29], c.Seq)
	if c.IsLast {
		out[29] = 1
	}
	copy(out[30:], c.Data)
	return out
}

func EncodeScreen(s ScreenFrame) []byte {
	payloadLen := screenFrameHeaderSize + len(s.Data)
	out := make([]byte, 5+payloadLen)
	out[0] = KindScreenFrame
	binary.BigEndian.PutUint32(out[1:5], uint32(payloadLen))
	binary.BigEndian.PutUint16(out[5:7], s.SessionID)
	binary.BigEndian.PutUint32(out[7:11], s.Seq)
	if s.IsKeyframe {
		out[11] = 1
	}
	binary.BigEndian.PutUint64(out[12:20], s.PtsMillis)
	copy(out[20:], s.Data)
	return out
}

func decodeChunk(payload []byte) (*ChunkFrame, error) {
	if len(payload) < chunkHeaderSize {
		return nil, ErrMalformedChunk
	}
	c := &ChunkFrame{}
	copy(c.FileID[:], payload[0:16])
	c.Seq = binary.BigEndian.Uint64(payload[16:24])
	c.IsLast = payload[24] != 0
	c.Data = append([]byte(nil), payload[chunkHeaderSize:]...)
	return c, nil
}

func decodeScreen(payload []byte) (*ScreenFrame, error) {
	if len(payload) < screenFrameHeaderSize {
		return nil, ErrMalformedScreen
	}
	s := &ScreenFrame{}
	s.SessionID = binary.BigEndian.Uint16(payload[0:2])
	s.Seq = binary.BigEndian.Uint32(payload[2:6])
	s.IsKeyframe = payload[6] != 0
	s.PtsMillis = binary.BigEndian.Uint64(payload[7:15])
	s.Data = append([]byte(nil), payload[screenFrameHeaderSize:]...)
	return s, nil
}

// FrameReader incrementally parses frames from arbitrary byte segments.
// Unknown kinds are skipped (forward compatibility), matching Swift.
type FrameReader struct {
	buf            []byte
	SkippedUnknown int
}

func (r *FrameReader) Append(data []byte) ([]Frame, error) {
	r.buf = append(r.buf, data...)
	var frames []Frame
	maxAllowed := MaxControlPayload
	if MaxChunkData+chunkHeaderSize > maxAllowed {
		maxAllowed = MaxChunkData + chunkHeaderSize
	}
	if MaxScreenData+screenFrameHeaderSize > maxAllowed {
		maxAllowed = MaxScreenData + screenFrameHeaderSize
	}
	for {
		if len(r.buf) < 5 {
			break
		}
		kind := r.buf[0]
		length := int(binary.BigEndian.Uint32(r.buf[1:5]))
		if length > maxAllowed {
			return frames, ErrOversizedFrame
		}
		if len(r.buf) < 5+length {
			break
		}
		payload := append([]byte(nil), r.buf[5:5+length]...)
		r.buf = r.buf[5+length:]

		// maxAllowed above is only the shared ceiling: it lets a file chunk ride
		// in at the screen frame's 4 MiB limit, twice the 2 MiB the protocol
		// documents for chunks. Hold each kind to its own cap (2026-08-17;
		// Swift and Kotlin do the same).
		switch kind {
		case KindControl:
			if length > MaxControlPayload {
				return frames, ErrOversizedFrame
			}
			frames = append(frames, Frame{Kind: kind, Control: payload})
		case KindFileChunk:
			if length > MaxChunkData+chunkHeaderSize {
				return frames, ErrOversizedFrame
			}
			c, err := decodeChunk(payload)
			if err != nil {
				return frames, err
			}
			frames = append(frames, Frame{Kind: kind, Chunk: c})
		case KindScreenFrame:
			if length > MaxScreenData+screenFrameHeaderSize {
				return frames, ErrOversizedFrame
			}
			s, err := decodeScreen(payload)
			if err != nil {
				return frames, err
			}
			frames = append(frames, Frame{Kind: kind, Screen: s})
		default:
			r.SkippedUnknown++
		}
	}
	return frames, nil
}
